### PR TITLE
ROS2 Dashing Image calibration

### PR DESCRIPTION
**Pull request by klintan**
Dusted off an old Ros1 --> Ros2 transformation of this one.

Would love some feedback on this :) 

@clydemcqueen feel free to try it out if you have the time.

### CODE REVIEW

**Comment by luca-della-vedova** [comment]
Hi Andreas (or should I say Hej!)

I'm trying to get this up and running on my side but am currently failing.
For now just two changes I needed to get it up and running:

- Create the resource/camera_calibration file and add the following to your package.py (patch format):
```
diff --git a/camera_calibration/setup.py b/camera_calibration/setup.py
index 274112b..2a7bdaf 100644
--- a/camera_calibration/setup.py
+++ b/camera_calibration/setup.py
@@ -7,11 +7,15 @@ setup(
     name=PACKAGE_NAME,
     version='1.12.23',
     packages=["camera_calibration", "camera_calibration.nodes"],
+    data_files=[
+    ('share/ament_index/resource_index/packages',
+      ['resource/' + PACKAGE_NAME]),
+    ('share/' + PACKAGE_NAME, ['package.xml']),
+    ],
     py_modules=[],
     package_dir={'': 'src'},
     install_requires=[
```
Without this change the package is not found by ros2 run, also this is the format suggested by the [example python packages](https://github.com/ros2/examples/blob/master/rclpy/topics/minimal_publisher/setup.py)

- Also I suppose you installed the opencv-python with pip? Even when installing the libraries the opencv-python dependency was not met on my system. The package actually worked removing it from the setup.py script, I took inspiration from the [vision_opencv package](https://github.com/ros-perception/vision_opencv/blob/ros2/cv_bridge/setup.py) which imports opencv but doesn't need to specify it as a dependency, although to be honest I'm not sure that is a good practice.

In the meanwhile I'll try to get some bags and try the new calibration node, from what I have seen it looks good.

Cheers!
/Luca

**Comment by klintan** [comment]
> Hi Andreas (or should I say Hej!)
> 
> I'm trying to get this up and running on my side but am currently failing.
> For now just two changes I needed to get it up and running:
> 
> * Create the resource/camera_calibration file and add the following to your package.py (patch format):
> 
> ```
> diff --git a/camera_calibration/setup.py b/camera_calibration/setup.py
> index 274112b..2a7bdaf 100644
> --- a/camera_calibration/setup.py
> +++ b/camera_calibration/setup.py
> @@ -7,11 +7,15 @@ setup(
>      name=PACKAGE_NAME,
>      version='1.12.23',
>      packages=["camera_calibration", "camera_calibration.nodes"],
> +    data_files=[
> +    ('share/ament_index/resource_index/packages',
> +      ['resource/' + PACKAGE_NAME]),
> +    ('share/' + PACKAGE_NAME, ['package.xml']),
> +    ],
>      py_modules=[],
>      package_dir={'': 'src'},
>      install_requires=[
> ```
> 
> Without this change the package is not found by ros2 run, also this is the format suggested by the [example python packages](https://github.com/ros2/examples/blob/master/rclpy/topics/minimal_publisher/setup.py)
> 
> * Also I suppose you installed the opencv-python with pip? Even when installing the libraries the opencv-python dependency was not met on my system. The package actually worked removing it from the setup.py script, I took inspiration from the [vision_opencv package](https://github.com/ros-perception/vision_opencv/blob/ros2/cv_bridge/setup.py) which imports opencv but doesn't need to specify it as a dependency, although to be honest I'm not sure that is a good practice.
> 
> In the meanwhile I'll try to get some bags and try the new calibration node, from what I have seen it looks good.
> 
> Cheers!
> /Luca

Haha hej! 
Awesome, thank you so much for testing it out and reviewing. Do you want to create a PR to my branch to add those ? If not I'll add it immediately.

**Comment by luca-della-vedova** [comment]
Hi Andres!

Sure I opened a pull request on your branch. I currently don't have a ROS2 camera and I get  calibration results which are way off when I use my ROS1 camera with a ros1->2 bridge, but again this might be due to the bridging.
Did you test if it works well for native ros2 cameras / bags?

/Luca

**Comment by klintan** [comment]
> Hi Andres!
> 
> Sure I opened a pull request on your branch. I currently don't have a ROS2 camera and I get calibration results which are way off when I use my ROS1 camera with a ros1->2 bridge, but again this might be due to the bridging.
> Did you test if it works well for native ros2 cameras / bags?
> 
> /Luca

Hm I tried it out and at least it looks like OK results when looking at the rectified image. However the tests fails and I don't really see what changes in the code that could ever make them fail (except for if maybe maybe the tests files are different?).  But I've been wrong before :D 

What was your results?

**Comment by klintan** [comment]
@luca-della-vedova could you elaborate on the issues you had?

**Comment by luca-della-vedova** [comment]
Hi Andres!

The main issue I had was that the calibration result was completely off, the image after calibration would be a very zoomed / distorted version of the original one, and all the parameters would be more than one order of magnitude from the real ones.
I suspected an issue in the data that is sent to the OpenCV function hence it returning bogus results but then got caught in other projects before I could figure it out!

/Luca

**Comment by luca-della-vedova** [comment]
Hi Andres,

A few updates since I had a bit of time to test it out, I can now "reproduce" that the rectified image looks alright, however my distortion parameters are completely off (which I can see also happens in the unit test results, where the distortion parameters are in the order of thousands!)

Can you confirm whether the camera matrix, as well as the rectified image, look OK in your camera calibration?

**Comment by luca-della-vedova** [comment]
Hi Andres,

I'm happy to report that I found the issue! It came up when migrating the code from Python 2 to Python 3, the code in calibrator.py line 338:
`opts_loc[j, 0, 0] = (j / b.n_cols)`
was relying on the behavior of the / operator between two integers in Python 2, which does integer division, but changed in Python 3 to return a floating point value. This resulted in the object points array to be wrong, which in turn made the results of the calibration completely bogus.

There was also a minor mistake during migration, for both of those I will open a pull request on your branch and I'm happy (!) to report that tests and calibration on my cameras are successful now (except some error due to vision_opencv package being outdated, currently addressed in https://github.com/ros-perception/vision_opencv/pull/295 ).

Can you try it out?

Cheers!
/Luca

**Comment by klintan** [comment]
> Hi Andres,
> 
> I'm happy to report that I found the issue! It came up when migrating the code from Python 2 to Python 3, the code in calibrator.py line 338:
> `opts_loc[j, 0, 0] = (j / b.n_cols)`
> was relying on the behavior of the / operator between two integers in Python 2, which does integer division, but changed in Python 3 to return a floating point value. This resulted in the object points array to be wrong, which in turn made the results of the calibration completely bogus.
> 
> There was also a minor mistake during migration, for both of those I will open a pull request on your branch and I'm happy (!) to report that tests and calibration on my cameras are successful now (except some error due to vision_opencv package being outdated, currently addressed in [ros-perception/vision_opencv#295](https://github.com/ros-perception/vision_opencv/pull/295) ).
> 
> Can you try it out?
> 
> Cheers!
> /Luca

Awesome, great job! and thanks for digging deeper into this. 

Looking forward to the PR:s and I'll test it out as soon as possible :)

**Comment by klintan** [comment]
Now the `colcon test` commands locally with the changes needed from here https://github.com/ros-perception/vision_opencv/pull/295
Until that is merged (or similar changes) it will continue to fail, but at least the results now seems sane with some awesome work put in by @luca-della-vedova . 

Luca -> I couldn't get the tests to run using colcon without the `CMakelists.txt` but seems you did, so perhaps I can just remove those? 

@SteveMacenski sorry to ask you for even more reviewing, but are you or someone else available for eventual review of this as well? Especially the `CMakeLists.txt` files both test and main, since I've never used them in a python project, and not really sure how they should look like or if they are even needed, I tried to do something similar to `image_geometry`, but might be a bit whack.

File: camera_calibration/package.xml
Comment (SteveMacenski) [inline camera_calibration/package.xml:11]: Readd maintainers. We need these emails.

File: camera_calibration/setup.py
Comment (SteveMacenski) [inline camera_calibration/setup.py]: Maintainers here too

File: camera_calibration/src/camera_calibration/calibrator.py
Comment (SteveMacenski) [inline camera_calibration/src/camera_calibration/calibrator.py:338]: Double slash?

File: camera_calibration/src/camera_calibration/nodes/cameracheck.py
Comment (SteveMacenski) [inline camera_calibration/src/camera_calibration/nodes/cameracheck.py]: You should init as the first thing where it was again

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
My 2 main concerns here are:

1) How does remapping / renaming work here for when they have different topic names? Before there was some resolving namespaces and global paths but I'm seeing alot of local paths here
2) Can you give me how you tested this?  
3) Tests are failing, not seeing how these failing tests relate to opencv_vision being broken

Also, add image calibration to the image pipeline metapackage package xml that's commented out

File: camera_calibration/package.xml
Comment (klintan) [reply camera_calibration/package.xml:11]: Ah of course, I've got no idea why I would remove you?

File: camera_calibration/src/camera_calibration/calibrator.py
Comment (klintan) [reply camera_calibration/src/camera_calibration/calibrator.py:338]: Makes sure it returns an int which was the default behavior in Py2 for a division (py3 will return a float, https://stackoverflow.com/questions/1535596/what-is-the-reason-for-having-in-python)

Comment (SteveMacenski) [reply camera_calibration/src/camera_calibration/calibrator.py:338]: Well I think that shows that it returns a floored float, ei `2.000` not `2`.

**Comment by JWhitleyWork** [comment]
@klintan - Please rebase on `ros2` branch and CI should be fixed.

**Comment by SteveMacenski** [comment]
@JWhitleyAStuff his issue is a test failure not a build failure - but yes rebasing would make life easier

File: camera_calibration/src/camera_calibration/calibrator.py
Comment (luca-della-vedova) [reply camera_calibration/src/camera_calibration/calibrator.py:338]: In any case it was going to be converted to float when assigned to the array, but the result of the right hand expression is now the same as the previous version with Python 2.

**Comment by luca-della-vedova** [comment]
> Luca -> I couldn't get the tests to run using colcon without the `CMakelists.txt` but seems you did, so perhaps I can just remove those?

Yea it still works alright if I check out my branch without the CMakelists, is it not running at all or is it failing? Did you try with my fork?

**Comment by klintan** [comment]
Pretty sure this will fail CI until this is merged https://github.com/ros-perception/vision_opencv/pull/295 (which I see @SteveMacenski already made a request for a time back)

~~I also made sure it passes Flake8 tests, let me know if you want me to revert that commit for readability, and put those changes in a different PR.~~ 
 
edit:  I'll hold off on the Flake8, is it possible to get the ROS2 flake8 configuration file somewhere? I saw that the line length was set to 120 instead of 80, but I was wondering maybe there are other changes as well?

**Comment by SteveMacenski** [comment]
No idea, I don't write much python.

**Comment by clydemcqueen** [comment]
Hey, @klintan -- sorry for the long delay. I tried out your PR on a few cheap USB cameras and compared the results from ROS1 to ROS2, and the results looks reasonable to me. Let me know if you need more help. Thanks!

**Comment by clydemcqueen** [comment]
Re: flake8 config, this looks promising: https://github.com/ament/ament_lint/blob/master/ament_flake8/ament_flake8/configuration/ament_flake8.ini

**Comment by SteveMacenski** [comment]
@klintan CI fails because of a failed test, not because of needing that PR. Is this actually blocked by anything? Please just make the tests pass and we're good.

@JWhitleyAStuff can you review, I'm OK with it. 

Has anyone tested this?

**Review by SteveMacenski — APPROVED** [review]

**Comment by JWhitleyWork** [comment]
@SteveMacenski - The test failure is not an issue with his code, it's an issue with the linter. Here is the full output when running tests on `camera_calibration`:

```
Starting >>> camera_calibration
[0.551s] ERROR:colcon.colcon_core.task.python.test:Exception in Python testing step extension 'pytest': attrib() got an unexpected keyword argument 'convert'
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/colcon_core/task/python/test/__init__.py", line 75, in test
    return await extension.step(self.context, env, setup_py_data)
  File "/usr/lib/python3/dist-packages/colcon_core/task/python/test/pytest.py", line 65, in step
    from pytest import __version__ as pytest_version
  File "/usr/lib/python3/dist-packages/pytest.py", line 13, in <module>
    from _pytest.fixtures import fixture, yield_fixture
  File "/usr/lib/python3/dist-packages/_pytest/fixtures.py", line 842, in <module>
    class FixtureFunctionMarker(object):
  File "/usr/lib/python3/dist-packages/_pytest/fixtures.py", line 844, in FixtureFunctionMarker
    params = attr.ib(convert=attr.converters.optional(tuple))
TypeError: attrib() got an unexpected keyword argument 'convert'

Failed   <<< camera_calibration	[ Exited with code 1 ]

Summary: 0 packages finished [0.46s]
  1 package failed: camera_calibration
```

The error is with the specific version of `pytest` and `attrs` that is currently available with 18.04 under the packages `python3-pytest` and `python3-attrs`. This can be verified by using `pip3` to upgrade `pytest` with: `pip3 install --upgrade pytest`. The tests will then pass.

@klintan - Please add the package `python3-pip` to the end of the `apt install` line in `.circleci/config.yml` and then, add the line `pip3 install --upgrade pytest` on the next line. This should solve the CI issue until a new version is released for 18.04.

**Comment by klintan** [comment]
> Re: flake8 config, this looks promising: https://github.com/ament/ament_lint/blob/master/ament_flake8/ament_flake8/configuration/ament_flake8.ini

Awesome thanks, I'll use that for my local config as well.

**Comment by SteveMacenski** [comment]
I think if you just run `ament_flake8` you should get it automatically

**Comment by klintan** [comment]
> Please add the package python3-pip to the end of the apt install line in .circleci/config.yml and then, add the line pip3 install --upgrade pytest on the next line. This should solve the CI issue until a new version is released for 18.04.

@JWhitleyAStuff tried this, but the tests still seems to fail, how did you get the full output, I think I'm missing something because the circle output I see is only 

```
Starting >>> camera_calibration
]0;colcon test [0/6 done] [1 ongoing]Starting >>> depth_image_proc
Failed   <<< camera_calibration	[ Exited with code 2 ]
]0;colcon test [0/6 done] [2 ongoing]]0;colcon test [1/6 done] [1 ongoing]Starting >>> image_proc
]0;colcon test [1/6 done] [2 ongoing]Finished <<< depth_image_proc [4.02s]
]0;colcon test [2/6 done] [1 ongoing]Starting >>> image_publisher
]0;colcon test [2/6 done] [2 ongoing]Finished <<< image_proc [2.57s]
]0;colcon test [3/6 done] [1 ongoing]Starting >>> image_rotate
]0;colcon test [3/6 done] [2 ongoing]Finished <<< image_publisher [2.64s]
]0;colcon test [4/6 done] [1 ongoing]Finished <<< image_rotate [2.79s]
]0;colcon test [5/6 done] [0 ongoing]Starting >>> image_pipeline
]0;colcon test [5/6 done] [1 ongoing]Finished <<< image_pipeline [0.03s]
]0;colcon test [6/6 done] [0 ongoing]
Summary: 5 packages finished [8.27s]
  1 package failed: camera_calibration
```


> I think if you just run `ament_flake8` you should get it automatically

I think I'll create a new PR for fixing all PEP8 issues to pass the ament_flake8 tests, otherwise it'll be a tad difficult to see the changes made (even if the review for this is done, I don't want something to sneak in when I do formatting changes)

**Comment by klintan** [comment]
Just wanted to circle back on this one, I've been a bit inactive, however trying again I still get failures and I don't really know how to the the full circle output? Could you give me some pointers here @JWhitleyAStuff ?
It all passes locally, I thought it might be permissions issues on circle, but even changing to `/tmp` did not solve it so I'm thinking i might have some dependency version mismatch locally.

**Comment by SteveMacenski** [comment]
Any progress / questions?

**Comment by klintan** [comment]
> Any progress / questions?

Haven't been able to track down this latest: 
```
AttributeError: 'CameraInfo' object has no attribute 'K'
```

It could be because of this https://github.com/ros-perception/vision_opencv/pull/295 and since it was merged just 9 days ago, i'm guessing it hasn't gotten it's way into a new release yet, meaning the rosdeps on circle will not pick it up. However I'm not sure that is the culprit :/

**Comment by JWhitleyWork** [comment]
@klintan - I believe I found the culprit: https://github.com/ros-perception/image_pipeline/blob/4d2b3398530f9033e88da945215a417930bfac53/camera_calibration/src/camera_calibration/calibrator.py#L438

For ROS2, a message variable can't start with an upper-case letter so these parameters were changed to `d, k, r`, and `p`. See https://github.com/ros2/common_interfaces/blob/d4505b57ed20c49853350705ea72eb6f89b2eafb/sensor_msgs/msg/CameraInfo.msg#L73

**Comment by klintan** [comment]
> @klintan - I believe I found the culprit:
> 
> https://github.com/ros-perception/image_pipeline/blob/4d2b3398530f9033e88da945215a417930bfac53/camera_calibration/src/camera_calibration/calibrator.py#L438
> 
> For ROS2, a message variable can't start with an upper-case letter so these parameters were changed to `d, k, r`, and `p`. See https://github.com/ros2/common_interfaces/blob/d4505b57ed20c49853350705ea72eb6f89b2eafb/sensor_msgs/msg/CameraInfo.msg#L73

Those are changed already in this PR though. 

https://github.com/ros-perception/image_pipeline/pull/447/files#diff-6b6b63c6b6a9ea7b1cfbd0c5dce79575R438-R441

**Comment by luca-della-vedova** [comment]
The image_geometry fix landed in the main eloquent packages (but not dashing yet).
Would it be acceptable to change the CI (from now on) to eloquent?

**Comment by SteveMacenski** [comment]
We’re targeting dashing or the most recent LTE until we can get a full stack ported release. 

Leaving dashing in a failing state isn’t ideal.

**Comment by maximus5684** [comment]
I'm not sure that I or @klintan knew what was going on here but it appears that @SteveMacenski and @luca-della-vedova may have had a conversation about it. Looks like https://github.com/ros-perception/vision_opencv/pull/295/files is the culprit which caused the `image_geometry.StereoCameraModel` to have a bug in the `fromCameraInfo` function. We are waiting on a Dashing package sync to release this fix into the public repo.

**Comment by SteveMacenski** [comment]
If it is possible to run CI **locally** with the tagged version of dashing of that package, build it all together, and tests pass, I'm willing to merge it. I want to know however there's a tagged version that its being tested against that will be released in the next or later dashing sync.

**Comment by JWhitleyWork** [comment]
@SteveMacenski - I was able to get CI to pass locally by adding:

`git clone https://github.com/ros-perception/vision_opencv.git -b 2.1.4 src/vision_opencv`

in the CircleCI config file. I can add this to CI if you want it to pass for the time being. It will increase the build time by a minute or more but will make the test pass here. You can verify that 2.1.4 is prepared to go into main from here: http://repo.ros2.org/status_page/ros_dashing_default.html?q=image_geometry.

I'm actually more worried about the `pytest` bug since it likely won't get upstreamed into the 18.04 repos so our CI is doing something that the build farm won't (upgrading `pytest`). We may have to investigate that one further to find a work-around in the code rather than my  temporary solution.

**Comment by SteveMacenski** [comment]
>  tactical commit, rerunning circleci 

Also got my attention again - sorry I saw this and then it got lost in the noise. I will take a look again on Monday and set a reminder for myself.

**Comment by klintan** [comment]
> > tactical commit, rerunning circleci
> 
> Also got my attention again - sorry I saw this and then it got lost in the noise. I will take a look again on Monday and set a reminder for myself.

Yeah thanks, I looked at the status and it seemed that 2.1.4 was now in dashing, but it seems I was incorrect, or at least the image dashing-core was not rebuilt/did not have it yet.

**Comment by JWhitleyWork** [comment]
@klintan - It's in `ros-shadow-fixed` but not in the public repo yet: http://repo.ros2.org/status_page/ros_dashing_default.html?q=image_geometry

The last Dashing sync was on 12/12/2020 so I expect to see another one any day now.

**Comment by SteveMacenski** [comment]
OK, we should wait for that sync. I don't want to break CI for the repo and then we just forget about it because other problems appear.

**Comment by luca-della-vedova** [comment]
[Sync is out](https://discourse.ros.org/t/new-packages-and-patch-release-for-ros-2-dashing-diademata-2020-03-19/13300)! Can we rerun CI on this?

**Comment by JWhitleyWork** [comment]
She be runnin!

**Comment by SteveMacenski** [comment]
@klintan everything is updated now -- its building!

The problem now is that it is failing the tests. Can you fix that test and then we can merge this finally :1st_place_medal:

**Comment by klintan** [comment]
> @klintan everything is updated now -- its building!
> 
> The problem now is that it is failing the tests. Can you fix that test and then we can merge this finally 🥇

Hm it seems  `python3-requests` is not installable using rosdep; which is kind of strange since it is available in the apt repo: `apt-cache show python3-requests`.

`python-requests` is available, but that is the py2 version.

Could you chime in @SteveMacenski do we need to do some workaround to get this working?

**Comment by JWhitleyWork** [comment]
Looks like https://github.com/ros/rosdistro/blob/7d569a60d55ab33c1520d62b1fe8894c42e533be/rosdep/python.yaml#L3711 needs some updates for Bionic.

@klintan If you make a pull request to the above file and add a `bionic` entry for `python3-requests` exactly like the others listed there, they should accept it. Once that is merged, CI should pass.

**Comment by klintan** [comment]
> Looks like https://github.com/ros/rosdistro/blob/7d569a60d55ab33c1520d62b1fe8894c42e533be/rosdep/python.yaml#L3711 needs some updates for Bionic.

thanks @JWhitleyWork. Yes seems bionic needs python3-requests added to it's rules for this to work.

**Comment by SteveMacenski** [comment]
I also think then its back to `python-requests` not `python3`? The python.yaml always confuses me by convention. Some have a different key for `python3-*` vs `python-*` and others are just in it for the OS that have python3 targeted

**Comment by klintan** [comment]
> I also think then its back to `python-requests` not `python3`? The python.yaml always confuses me by convention. Some have a different key for `python3-*` vs `python-*` and others are just in it for the OS that have python3 targeted

Yeah I tried the `python-requests` as well, and that was the py2 version it seems.

**Comment by SteveMacenski** [comment]
@klintan I'm not sure what's happening in this particular entry: https://github.com/ros/rosdistro/blob/7d569a60d55ab33c1520d62b1fe8894c42e533be/rosdep/python.yaml#L3711

I've never seen the `'*':` before and then `<distro>_python3` as sub-entries, this is odd. Usually there are `python-<library>` entries with `<distro>` keys and then a separate `python3-<library>` with the `<distro>` keys. I looked over the pattern through and it looks like you just need to add a 

```
    bionic_python3: [python3-requests]
```
 
to that entry. Personally, that looks super unmaintainable, we really should move all those to `python3-<library>` so that then a `'*'` covers all future distros unless something changes. But that's a problem for another day I suppose. It looks to be 77 instances of that pattern.

**Comment by klintan** [comment]
> bionic_python3: [python3-requests]

Thanks for you patience and explanation. Here is the PR https://github.com/ros/rosdistro/pull/24292

**Comment by SteveMacenski** [comment]
@klintan let us know when its merged and we can re-run CI and hopefully be done with this :-)

**Comment by klintan** [comment]
> @klintan let us know when its merged and we can re-run CI and hopefully be done with this :-)

Merged now 👍 @SteveMacenski

**Comment by JWhitleyWork** [comment]
> Merged now  @SteveMacenski

CI has been kicked! Let's see what she does.

**Comment by JWhitleyWork** [comment]
And it's gooooooooood! :tada:

**Comment by SteveMacenski** [comment]
Yeah!! Thanks @klintan almost a year later...

**Comment by SteveMacenski** [comment]
@JWhitleyWork Am I correct in saying that we're done with the ROS2 initial port now?

**Comment by JWhitleyWork** [comment]
> @JWhitleyWork Am I correct in saying that we're done with the ROS2 initial port now?

I hesitate to say it but I think so! There are no more COLCON_IGNORE or AMENT_IGNORE files in the stack so I think it's all building! The TODO list is pretty long but I think we can actually release something now.

**Comment by klintan** [comment]
> @JWhitleyWork Am I correct in saying that we're done with the ROS2 initial port now?

One small thing;  I couldn't run `image_proc` composable nodes from the launch file (PR #457 ). But maybe that is a bug you could live with until fixed the things in that PR ? https://github.com/ros-perception/image_pipeline/pull/457 (which is also like 8 months old unfortunately). Took a look and seems that hasn't been fixed/updated in the `image_proc` code in another PR either.
However a bunch of changes needed for that one before merge :/ I'll try to get them done asap.

**Comment by SteveMacenski** [comment]
Does #457 also cover #468? Would love to see that in. That's the last of the "make things work" tickets. 

I think that then lets us focus on the Foxy release, clean up and clear up the code from 10 years of community PRs a little, and complete this port. @JWhitleyWork is there any vision for this project you want to mention other than a straight port? I don't know in what context  / if alot of people use this. If so, perhaps there's some tangible improvements we can make (expose new openCV functionality that fits in this scope introduced in the last 5 years, add new calibration standards based on kalibr, visual inertial or extrinsic calibration, etc)

**Pull request merged**
